### PR TITLE
Random album shuffling

### PR DESCRIPTION
**Pull request by Octoton**
**This MR is a work in progress"**
 
I like to listen to album randomly (and not just song) and I don't seems to be the only one (see https://github.com/AdrienPoupa/VinylMusicPlayer/issues/255 and https://github.com/AdrienPoupa/VinylMusicPlayer/issues/233)

This is done by adding a new shuffling mode (disabled by default as icon can be confusing when you don't know what it do:
Screenshot_20201011-175855_Vinyl_DEBUG

Moreover when activated, a new song will be added at the end of the playing queue to show which random album was chosen to be played next. The menu of this album is there to allow you to change the next album in three different mode:
 - randomly
 - following the last song of the queue genre
 - following the last song of the queue artist

Screenshot_20201011-175902_Vinyl_DEBUG

There is a listen album history and a searched next album history so that the shuffle mode will not always pick the same album over and over again (limited to a depth of from 0 to 9 album as it is just there for better user experience)

In preferences (at the end of "Now playing"), you can choose the depth of those history and the criteria used to search the next ablum (by genre, artist or randomly with fallback option if no other album outside the listen history is found following those criteria)

### CODE REVIEW

**Comment by Octoton** [comment]
@AdrienPoupa this is a new feature that i think will be really useful for some people and could be extended in the futur

**Comment by Octoton** [comment]
@AdrienPoupa @soncaokim there is a new functionality for this app: album shuffling
I have finish implementing everything I wanted on this, could you tell me what you think of it, the usability/bug problem you found, ect...

**Comment by soncaokim** [comment]
If I get your idea, we'd have a perpetual endless playing queue that gets filled with new songs (new album in your case). 
When playback reaches the end of queue, it will be fed with extra items.

If yes, this is interesting and could be extended to have the feeding source to be others (random album, specific genre, songs from specific artist, a search query, a dynamic playlist).

Is that compatible/conceivable with your proposition?

**Comment by Octoton** [comment]
It could be possible in theory but the all search system is center on finding other album, I work with list of Album.
And I am not sure how to give the user option that are not overwhelming/are easy to grasp.

For more context:
at the present when the queue is loaded, I look at the last song characteristic and from the preference set by the user select a new album with the same characteristic, so either same artist, same genre, random.
For example in the screenshot below the user as chosen to first try to find a new album with the same Artist and if no other album exist (or it as already been listen too before) the system will choose randomly. But the user could have set this up to find other album of the same Genre and if nothing is found don't propose a next album
Screenshot_20210719-220048
  
When you click on a criteria you get to chose from a list of option:
Screenshot_20210719-220053

You can have from 1 to 3 fallback criteria and as you add more, you have less and less option.
For the first: "Artist", "Genre", "Random"
Second only show if artist or genre is selected and will show "Genre or Artist" (show only the one remaining), "Random", "None" (if you don't want a fallback option)
The third only is Genre or Artist is selected and will show "Random" or "None"
  

Finally if the proposed net album is not to your linking, you can click on the playlist 3-dot menu to change with the criteria that you want:
Screenshot_20210719-220112

**Comment by Octoton** [comment]
I will look into it a little to see find something that make sense, and is not feature creep

**Comment by soncaokim** [comment]
Some general comments:
- Can you please layout the main design ideas, and the purpose of different classes/modules that you plan to introduce? I find it hard to follow and the changes are too intrusive to me (many change that I cannot justify).
- Is it possible to separate the new implementation in different parts, with clear and minimal interface between them? I'm thinking of:
  - an auto-feeding playing queue, with necessary support from music service. I guess that you might need to introduce some UI change as well to support this, but I'd prefer to keep it as least intrusive to the existing classes as possible.
  For instance, `PlayingQueueAdapter` and `Discography` should not need to know that there are `AlbumShuffling` options.
  - an abstract feeding system, for instance:
    ```
    interface PlayingQueueFeeder {
       List<Song> nextBatch(Bundle criteria);
    }
    ```
  - one reference implementation for feeding by album (your idea), plus necessary UI and preferences support.
  - instead of creating a new DB for storing album playback history, why cant you rely on the existing History.db?

These are major changes to the proposal, but I think this will help adoption easier.

**Comment by Octoton** [comment]
1/ The main idea of this system is that at the end of the playing queue, `MusicService` will add a new song which as an incorrect id: NEXT_RANDOM_ALBUM_SONG_ID and information about which album to load (with the help of `NextRandomAlbum`).
Thus when the `PlaybackHandler` see that the song it is trying to load is the next album it will load it.
EMPTY_NEXT_RANDOM_ALBUM_ID is there to easily see that `NextRandomAlbum` didn't found any new album to load (and continue to show something to the user).
This invalid song is then shown to the user like this:
image
`MusicUtil` modification ensure that we don't try to show the song number and `PlayingQueueAdapter` line 85 replace the position number with a "-" to better differentiate this from the other song.
Finally in `PlayingQueueAdapter`, I need to ensure that the context menu of this song will be different from the other menu as a lot of option don't make sense there, and new option need to be added, (see below the specific menu)
image

`Discography` modification was a way to ensure that this fake song would never be added to the cache (but perhaps this is not needed as the song id is invalid).
  
  
All new class are set in misc/AlbumShuffling:
there is `NextRandomAlbum` that is the center of it all, it initialize the search and give a result
`AutomaticSearch` is there to give the search implementation used when a new playlist queue is loaded (behavior can be modified in preference)
'ManualSearch' is there to give the search implementation when the user use the context menu (if they don't like the `AutomaticSearch` result)
`History` is there to remember the last album that where listen too (and not all of the song that are inside of it), it help to ensure that the next album is not always the same
`DB` is there to ensure listenHistory (in `NextRandomAlbum`) is not forgotten every time the app idle for too long, moreover it remember the invalid song to add a the end of the playingQueue (because the song id is invalid, `MusicPlaybackQueueStore` will not be able to get it back as only song id is remember in it and not album id or name)
  
  
Finally to as this is just a shuffling option, a new state was created in `MusicService`: SHUFFLE_MODE_SHUFFLE_ALBUM
with a new icon to go with it:
image

**Comment by Octoton** [comment]
3/ For the new history, it is necessary as I don't have any other option to get back the next album to show in the playingqueue (because the song id is invalid, MusicPlaybackQueueStore will not be able to get it back as only song id is remember in it and not album id or name)
and it is a convenience thing: as there is at max 9 albums than can be remember don't really need more as the history is only used to ensure that `NextRandomAlbum` will not always give the same result, it is quicker to directly remember the album as it is to try to recreate what album where listen to by looking at a list of song (and less prone to mistake).

**Comment by Octoton** [comment]
2/ I like what you are going for, will do some mockup with better abstraction level

**Comment by Octoton** [comment]
> If yes, this is interesting and could be extended to have the feeding source to be others (random album, specific genre, songs from specific artist, a search query, a dynamic playlist).

So, the short answer is that my implementation completely revolve on an album source so it will not work as is.
However, I am unsure about how to extend this concept:
if instead of choosing a next album, the system would load a next list of song I am not sure exactly which list of song to authorize as each new type of list would need to be able to be easily shown to the user as the last song of the playlist queue, be able to have criteria to jump from one list to another, have preference setting to better tune the system that make sense. 
  
Album list of song are easy as criteria are genre and artist, and preference is already one :)
Playlist could be done as the same criteria can be used (but new system need to be created to get the list of genre and list of artist in each playlist quickly | same system could be added into album as for now i only look at the artist and genre of the first and last song of any album)
Genre could be done but only criteria is Artist and I am ensure of how a user would like to use if (lot of the time a genre is already a big list of song and going from genre to genre could be pretty strange and user would never be able to guess which genre would come next), moreover, Album with search criteria == Genre is really close
Songs from specific artist is the same things a Album with search criteria == Artist
For a search query, it is really difficult to show what is next to the user, and the preference would be really difficult to design without being a mess

**Comment by soncaokim** [comment]
Let's target first the issue where core classes (MusicService, Discog, Preferences, playqueue UI classes) contain references to a specific dynamic queue implementation - a kind of dependency that I'd rather avoid.

If I understand correctly - please correct me if I'm wrong - the reasons are:
- you need to introduce the 'last element' UI artifact to indicate  the current feeding mechanism (next album) and related controlling options.
- when this element is reached, you need to trigger the fetching of next batch (next album)

To avoid that, how about:
1. isolate the play queue mgmt to a separated class, called StaticPlayQueue. Now it is embedded/entangled inside MusicService.
2. create a new DynamicPlayQueue class that contains the semantic of a special last item, as well support feeding.
  MusicService  only knows of either a static or dyn play queue. Or even only an AbstractQueue if we managed to make an abstraction.
4. UI for playqueue: as before, unless we are dealing with the last element of a dynqueue.
  by based on the position in queue instead of the type of element, we evade the need of introducing a special type of Song as well as change in discog. Note that there is already some similar position-based logic for the static queue, so the it would be more maintainable.
  If that special element is detected, delegate the UI decoration, menu handling, etc to a concrete implementation class.
5. move all other changes you proposed in a new package - eg dynqueue.byalbum.XXX.
  This contains the concrete implem of the playqueue UI for the last element, the UI to choose the different criteria and persistence support (via prefs or via DB).

6 and beyond. Add other kind of feeding implem (ideas I suggested before).

If this makes senses to you, the plan can be implemented in multiple steps/multiple PRs. I can offer some help on the first steps as well.

As said at the beginning, please correct/amend if I overlooked something.

And all this "plan sur la comète" makes sense if @AdrienPoupa is willing to adopt it and merge, so please give your voice as well.

**Comment by Octoton** [comment]
Fine by me for this type of abstraction,
The only problem that I have is how to stop PlayingQueueAdapter to work on song only (this is why i made my implementation like that), I don't have a good grasp of how this part work.
Seems like playingQueue only work on position and a list of song as it extend SongAdapter, thus I don't see how it can detect a special element without this element being a song ? And if we want to stop this adapter to look at list of song, it cannot just extend the Song Adapter which would mean a lot of this to re-implement on it.
Seems like I don't quite understand what you are going for in 3.

**Comment by soncaokim** [comment]
> how it can detect a special element without this element being a song ?
> Seems like I don't quite understand what you are going for in 3.

To be checked, but it seems to me that the way the "Shuffle All" button is implemented in the Library Songs tab is similar to what I'm trying to explain:
- the underlying dataset is the list of songs, but with position shifted by 1 to accommodate the special first element.
- the concept of this first element is contained inside the Songs tab.
- the UI for this special first is different from usual SongAdapter: no menu, special icon, special text and coloring, special action handler.

**Comment by Octoton** [comment]
Oh, I see
So it is the same concept as AbsOffsetSongAdapter.java but instead of putting the offset element in first, I can but it in last.
I think I will do 1. in a separated PR next week (as it just cleaner code)
And all the other point in the same PR  after the first one (modify only relevant if for album shuffling and beyond) but on different batch of commits so it is easier to talk about/modify those new implementation

Thus this PR is more likely to be closed in the week to come

**Comment by soncaokim** [comment]
Thanks!

One more thing as per how this new dynqueue is started by the user.

As per your proposal, I understand that you enable it via prefs, correct? So if the user needs to get there to enable/disable the feature.

With step 2 or 3, how about adding a new menu item (eg "Start new perpetual queue", the wording to be reviewed) next to the available menu items (beside "Play next" and "Add to queue"). This would be offered for Album to start with your proposal, but we can extend it to other element (song, artist, etc) later.

While browsing, If user chooses this choice , the UI to refine the criteria pops up to collect feeding criteria, then MusicService runs with the dynqueue instead of static queue.

If it makes sense, we can move the new Shuffle icon to the special last item as well (to avoid altering the core UI logic).

**Comment by Octoton** [comment]
In my proposal the settings part is more of a set and forget type of thing, and the new icon is the on/off button. See this video for when the user set this for the first time:

https://user-images.githubusercontent.com/56130419/126908136-9dcf6da2-b4cb-44eb-85b0-3bf53725dcc9.mp4

And then for all next use, only one click is needed:

https://user-images.githubusercontent.com/56130419/126908152-c8974691-116e-4fd4-af4f-0d7a728667ff.mp4

If instead of the new shuffle icon, we use a menu item, then to stop this we could put in the special last item menu a stop perpetual queue. However, i am not completely sure where to put the start item, on each song menu, only on the playingqueue song menu, or on the playingqueue global menu ? And each time the user click on it a fragment appear on screen to choose criteria ? Thus we don't need to have a settings page ?

**Comment by soncaokim** [comment]
Thanks for the videos.

> If instead of the new shuffle icon, we use a menu item, then to stop this we could put in the special last item menu a stop perpetual queue. 

Nice idea

> However, i am not completely sure where to put the start item, on each song menu, only on the playingqueue song menu, or on the playingqueue global menu ? 

What I proposed above is to add one more item in `R.menu.menu_album_detail` (there are many of them, one per each kind of object - song album etc - that we manipulate). The extra menu item goes together with "Play next" and "Add to queue" ones. However, not sure that this will make sense with the idea you propose - i.e. shuffle by album.

Guess we can start from playing queue global menu, and refine the implementation after.

> And each time the user click on it a fragment appear on screen to choose criteria ? Thus we don't need to have a settings page ?

Yes

**Comment by soncaokim** [comment]
> Guess we can start from playing queue global menu, and refine the implementation after.

Or we can start from the main three dots menu: shuffle all, new perpetual queue..., new playlist

Screenshot_20210725-214802~2

**Comment by Octoton** [comment]
As a purpose solution, seems like it would make sense in three place:
 - In the global menu on the tab Albums replacing Shuffle all, so each tab as a logical shuffle option in the futur: shuffle songs, shuffle albums, shuffle artist, shuffle genre, shuffle playlist
 - In the album detail menu close to play next (genre detail menu could get its own, artist too, ...).
 - In the playing queue global menu, something like: create a perpetual queue (the user can then chose  if he want by song/album or in the futur artist/genre/playlist)

**Comment by soncaokim** [comment]
@Octoton have you had time to push the ideas further? Is there anything that I can help with?

**Comment by Octoton** [comment]
Good timing, i am creating the first CR right now :)
